### PR TITLE
[WIP][SPARK-33444][ML] Added support for Initial model in Gaussian Mixture Model in ML

### PR DESCRIPTION
**Pull request by shahar603**
### What changes were proposed in this pull request?

Added an optional `initialModel` in the GaussianMixture class.

### Why are the changes needed?

To allow for non random initialization in ML's Gaussian Mixture. The algorithm's results can be affected due to the initialization and the existing random initialization may be limiting in some cases. This feature exists in MLLIB. 

### Does this PR introduce _any_ user-facing change?

* A new method: `setInitialModel` in the `GaussianMixture` class

### How was this patch tested?

Yes, tests were added to the `GaussianMixtureSuite.scala` file.
The tests test:
* The ability to add the model
* The fact the GaussianMixture object fit data using the initial parameters
* That the GaussianMixture class validates that `K` is the same to the initial model's `K`

### Why is this WIP?

Not sure if and how to add the `@Since` annotation (even after I've read the _How to contribute_ page). Also, I'm not sure how to add support for Python and R.

### CODE REVIEW

**Comment by shahar603** [comment]
The reason the tests fails is due `Param[GaussianMixtureModel]`. GaussianMixtureModel cannot be serializable using `jsonEncoder`.

mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala

So when the code tries to save/load the model, it fails because it can't serialize the GaussianMixtureModel object. Is there a way to fix this without changing the implementation of`jsonEncoder`? I believe changing `jsonEncoder` is beyond the scope of this PR and might require a new issue. What do you suggest?

File: mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala [hunk 1]
```diff
@@ -424,11 +463,11 @@ class GaussianMixture @Since("2.0.0") (
   }
 
   private def trainImpl(
-      instances: RDD[(Vector, Double)],
-      weights: Array[Double],
-      gaussians: Array[(DenseVector, DenseVector)],
-      numFeatures: Int,
-      instr: Instrumentation): (Double, Int) = {
+                         instances: RDD[(Vector, Double)],
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala:466]: Please revert all the whitespace changes first

**Comment by zhengruifeng** [comment]
in the .ml side, models do not support initialization for now. I think we need more discussion and design a general method to support them.

**Comment by shahar603** [comment]
> in the .ml side, models do not support initialization for now. I think we need more discussion and design a general method to support them.

Do you think this fits the scope of this PR? Should I close it because it shouldn't be merged until such changes are implemented which in my understanding is TBD.

**Pull request closed without merging**